github/commitstatus.go

```diff
@@ -96,14 +102,44 @@ func (s CommitStatus) Add(sha, state, targetURL, description string) error {
 	req.Header.Set("Accept", "application/vnd.github.v3+json")
 	req.Header.Set("Content-Type", "application/json")
 
+	for attempt := 0; attempt < maxRetries; attempt++ {
+		s.log.Info(fmt.Sprintf("Http request attempt: %d out of %d", attempt+1, maxRetries))
+		resp, err := httpRequestDo(req)
+		if err != nil {
+			return err
+		}
+
+		retry, timeToSleep := checkForRetry(resp, attempt)
+		if !retry {
+			return s.checkStatus(resp, state, sha, url)
+		}
+		time.Sleep(timeToSleep)
+	}
+	return nil
+}
+
+type httpResponse struct {
+	statusCode         int
+	body               string
+	oauthInfo          string
+	rateLimitRemaining int
+	rateLimitReset     time.Time
+}
+
+func httpRequestDo(req *http.Request) (httpResponse, error) {
```

Comment:
does the actual request, closes the response body (important), collects and stores all the data that would be required in subsequent steps...